File: elixir/anagram/anagram.exs
defmodule Anagram do

  import String, only: [ downcase: 1, graphemes: 1, to_atom: 1 ]
  import Enum, only: [ join: 1, sort: 1 ]

  @doc """
  Returns all candidates that are anagrams of, but not equal to, 'base'.
  """
  @spec match(String.t, [String.t]) :: [String.t]
  def match(base, candidates) do
    encoded_words(candidates, base) |> Keyword.get_values(encoded_word(base))
  end

  def encoded_words(words, base) do
    for word <- words, downcase(base) != downcase(word), into: Keyword.new, do: { encoded_word(word), word }
  end

  defp encoded_word(word) do
    word |> downcase |> graphemes |> sort |> join |> to_atom
  end

end

File: go/anagram/anagram.go
// Package anagram creates lists of possible anagrams
package anagram

import (
	"sort"
	"strings"
)

// Detect returns a correct list of anagrams, given a word and a list of possible anagrams
func Detect(subject string, candidates []string) (result []string) {
	var i int
	result = make([]string, len(candidates))
	encoded := encode(subject)
	for _, word := range candidates {
		if strings.ToLower(subject) != strings.ToLower(word) && encoded == encode(word) {
			result[i] = word
			i++
		}
	}
	return result[:i]
}

func encode(word string) string {
	word = strings.ToLower(word)
	letters := strings.Split(word, "")
	sort.Strings(letters)
	return strings.Join(letters, "")
}
